nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java

```diff
@@ -0,0 +1,286 @@
+                    getLogger().warn(result.getValue());
+                }
+            }
+
+            schema =
+                    schemaProperty != null
+                            ? new ObjectMapper().readValue(schemaProperty, Schema.class)
+                            : convertSchema(recordReader.getSchema());
+
+            List<String> fieldNames = schema.getFieldNames();
+
+            needInitFormatter = schema.getTimeType() != Schema.TimeType.LONG;
+
+            tablets = generateTablets(schema, maxRowNumber);
+            SimpleDateFormat timeFormatter = null;
+
+            Record record;
+
+            while ((record = recordReader.nextRecord()) != null) {
+                Object[] values = record.getValues();
+                if (timeFormatter == null && needInitFormatter) {
+                    timeFormatter = initFormatter((String) values[0]);
+                    if (timeFormatter == null) {
+                        getLogger().error("The format of time is not supported.");
+                        inputStream.close();
+                        recordReader.close();
+                        processSession.transfer(flowFile, REL_FAILURE);
+                        return;
+                    }
+                }
+
+                long timestamp;
+                if (needInitFormatter) {
+                    timestamp = timeFormatter.parse((String) values[0]).getTime();
+                } else {
+                    timestamp = (Long) values[0];
+                }
+
+                boolean flag = false;
+
+                for (Map.Entry<String, Tablet> entry : tablets.entrySet()) {
+                    String device = entry.getKey();
+                    Tablet tablet = entry.getValue();
+                    int rowIndex = tablet.rowSize++;
+
+                    tablet.addTimestamp(rowIndex, timestamp);
+                    List<MeasurementSchema> measurements = tablet.getSchemas();
+                    for (MeasurementSchema measurement : measurements) {
+                        String tsName =
+                                new StringBuilder()
+                                        .append(device)
+                                        .append(".")
+                                        .append(measurement.getMeasurementId())
+                                        .toString();
+                        int valueIndex = fieldNames.indexOf(tsName) + 1;
+                        Object value;
+                        try {
+                            TSDataType type = measurement.getType();
+                            value = convertType(values[valueIndex], type);
+                        } catch (Exception e) {
+                            value = null;
+                        }
+                        tablet.addValue(measurement.getMeasurementId(), rowIndex, value);
+                    }
+                    flag = tablet.rowSize == tablet.getMaxRowNumber();
+                }
```

Comment:
It would be helpful to refactor this method into smaller component parts to avoid deeply-nested loops.